src/backend/distributed/executor/multi_utility.c

```diff
@@ -3774,10 +3842,20 @@ ProcessDropTableStmt(DropStmt *dropTableStatement)
 
 		Oid relationId = RangeVarGetRelid(tableRangeVar, AccessShareLock, missingOK);
 
-		if (relationId == InvalidOid ||
-			!IsDistributedTable(relationId) ||
-			!ShouldSyncTableMetadata(relationId) ||
-			!PartitionedTable(relationId))
+		/* we're not interested in non-valid, non-distributed relations */
+		if (relationId == InvalidOid || !IsDistributedTable(relationId))
+		{
+			continue;
+		}
+
+		/* invalidate foreign key cache if the table involved in any foreign key */
+		if ((TableReferenced(relationId) || TableReferencing(relationId)))
+		{
+			shouldInvalidateForeignKeyGraph = true;
```

Comment:
I cannot actually see any edge cases with that. Either dropping the referenced or referencing table would invalidate the foreign key graph (depending on the order of the DROP).

What kind of edge do you foresee with that? @velioglu will add a test with the following specs:

```
-- create fkeys in the following form
schema_1.table_1 -> schema_2.table_1 ->schema_1.table_2

-- test1
drop schema schema_2;
check graph

-- test2
drop schema schema_1;
check graph
 
```

Reply:
Ok, we hit an issue, still looking into that :)

Reply:
An alternative might be to invalidate from `master_drop_distributed_table_metadata` . A drawback is that you cannot know whether there were any foreign keys once you're in the drop handler, so you'd probably always invalidate when a distributed table or reference table is dropped.

Reply:
I have added support for invalidating graph when a drop cascades to a table by processing the DROP SCHEMA command in multi_ProcessUtility.

Reply:
Hmm, I see what you mean. I think we missed that again :/ 

But is it ever possible for any command (except for `DROP SCHEMA` and `DROP DATABASE`) to drop a table? I cannot find any.

I think, say a `DROP TYPE`, can drop a column which is involved in the foreign key, which breaks our logic a bit. Or, `DROP INDEX` on the pkey could `CASCADE` to the foreign key.
Those sound like the only place we're missing the invalidation.

I'm actually not too worried about `DROP` commands because any lack of invalidation could only lead to enforcing some unnecessary restrictions until the graph is invalidated.

And, having support for `ALTER TABLE ADD/DROP fkey`, `CREATE/DROP TABLE with fkey` and `DROP SCHEMA` sounds like covering majority of the cases where invalidation is necessary.

We could maybe quickly add `DROP INDEX CASCADE` support as well cc: @velioglu.

@marcocitus Do you prefer/expect to have a solution with a fundamental change in the logic?

Reply:
In the mean time, I'll try to write a function that recursively goes trought `pg_depend` on any drop command, and if it finds any foreign key constraint being dropped, invalidates the foreign key.

Actually, almost copy & paste `findDependentObjects()` function in Postgres

Reply:
Or, given that it is not a simple function to write, I'll probably prefer to open an issue on that for now and leave it for future :/

Reply:
Yeh, I think we can punt on it a bit, since these are unlikely edge cases. 

citus_drop_trigger also gets called when dropping a constraint, which is something we may or may not be able to use.

Reply:
Opened an issue #2250